src/Cdn77/Sniffs/OrderingConventions/AlphabeticallyOrderedConstantsSniff.php

```diff
@@ -0,0 +1,258 @@
+{
+    public const CodeIncorrectConstantOrder = 'IncorrectConstantOrder';
+
+    public function register(): array
+    {
+        return [T_OPEN_TAG];
+    }
+
+    public function process(File $phpcsFile, mixed $stackPtr): void
+    {
+        $namesWithValuesByVisibility = $this->findConstantNamesWithValuesByVisibility($phpcsFile);
+
+        if ($namesWithValuesByVisibility === []) {
+            return;
+        }
+
+        foreach ($namesWithValuesByVisibility as $visibility => $namesWithValues) {
+            $constantNames = array_map(
+                static fn (array $nameWithValue): string => $nameWithValue['name']['content'],
+                $namesWithValues,
+            );
+            $sortedConstantNames = $constantNames;
+            sort($sortedConstantNames);
+
+            if ($sortedConstantNames === $constantNames) {
+                continue;
+            }
+
+            $firstNameWithValue = reset($namesWithValues);
+            assert(is_array($firstNameWithValue));
+
+            $fix = $phpcsFile->addFixableError(
+                sprintf('%s constant names are not alphabetically ordered.', ucfirst($visibility)),
+                $firstNameWithValue['name']['ptr'],
+                self::CodeIncorrectConstantOrder,
+            );
+
+            if (! $fix) {
+                continue;
+            }
+
+            $this->fix($phpcsFile, $namesWithValues);
+        }
+    }
+
+    /** @param list<NameWithValueShape> $namesWithValues */
+    private function fix(File $file, array $namesWithValues): void
+    {
+        $fixer = $file->fixer;
+        $sortedNameAndValueTokens = $namesWithValues;
+        usort(
+            $sortedNameAndValueTokens,
+            static fn (array $a, array $b): int => $a['name']['content'] <=> $b['name']['content'],
+        );
+
+        $fixer->beginChangeset();
+
+        foreach ($namesWithValues as $key => $nameWithValue) {
+            $sortedNameAndValueToken = $sortedNameAndValueTokens[$key];
+
+            $namePointer = $nameWithValue['name']['ptr'];
+            FixerHelper::removeBetweenIncluding($file, $namePointer, $namePointer);
+            $fixer->addContent($namePointer, $sortedNameAndValueToken['name']['content']);
+
+            $value = $nameWithValue['value'];
+            FixerHelper::removeBetweenIncluding($file, $value['startPtr'], $value['endPtr']);
+            $fixer->addContent($value['startPtr'], $sortedNameAndValueToken['value']['content']);
+        }
+
+        $fixer->endChangeset();
+    }
+
+    /** @return array<string, list<NameWithValueShape>> */
+    private function findConstantNamesWithValuesByVisibility(File $phpcsFile): array
+    {
+        $constantNamesWithValues = [];
+        $tokens = $phpcsFile->getTokens();
+
+        foreach ($tokens as $stackPtr => $token) {
+            if ($token['code'] !== T_CONST) {
+                continue;
+            }
+
+            $visibility = $this->getVisibility($phpcsFile, $stackPtr);
```

Comment:
nevim, jestli je ok zvazovat tu visibility. obcas chces mit kontanty zamichane dohromady nezavisle na visibility. protoze public mas ty, co pouzivas jinde. a private kdyz je nikde nepouzivas (aby ti to pak treba SA neoznacila za unused code)

Reply:
V cs pravidlech máme pravidlo SlevomatCodingStandard.Classes.ClassStructure, které vynucuje aby spolu byly grouped public, protected a private konstanty. Pokud budu visibility ignorovat a smíchám je, tak tohle pravidlo poruším. Asi jedině to udělat konfigurovatelné, ale nevím jestli je to right now worth it.

Reply:
Aha, nepříjemné. Nvm